packages/expo-updates/android/src/main/java/expo/modules/updates/errorrecovery/ErrorRecoveryHandler.kt

```diff
@@ -0,0 +1,137 @@
+    // remove any extraneous tasks from the pipeline as such
+    pipeline.retainAll(setOf(Task.WAIT_FOR_REMOTE_UPDATE, Task.CRASH))
+    delegate.markSuccessfulLaunchForLaunchedUpdate()
+  }
+
+  private fun handleRemoteLoadStatusChanged(newStatus: ErrorRecoveryDelegate.RemoteLoadStatus) {
+    if (!isWaitingForRemoteUpdate) {
+      return
+    }
+    isWaitingForRemoteUpdate = false
+    if (newStatus != ErrorRecoveryDelegate.RemoteLoadStatus.NEW_UPDATE_LOADED) {
+      pipeline.remove(Task.LAUNCH_NEW_UPDATE)
+    }
+    runNextTask()
+  }
+
+  private fun runNextTask() {
+    when (val nextTask = pipeline.removeAt(0)) {
+      Task.WAIT_FOR_REMOTE_UPDATE -> waitForRemoteUpdate()
+      Task.LAUNCH_NEW_UPDATE -> tryRelaunchFromCache() // only called after a new update is downloaded and added to the cache, so the implementation is equivalent
+      Task.LAUNCH_CACHED_UPDATE -> tryRelaunchFromCache()
+      Task.CRASH -> crash()
+      else -> throw RuntimeException("ErrorRecoveryHandler cannot perform task $nextTask")
+    }
+  }
+
+  private fun maybeStartPipeline(exception: Exception) {
+    encounteredErrors.add(exception)
+    if (delegate.getLaunchedUpdateSuccessfulLaunchCount() > 0) {
+      pipeline.remove(Task.LAUNCH_CACHED_UPDATE)
+    } else if (!hasContentAppeared) {
+      delegate.markFailedLaunchForLaunchedUpdate()
+    }
+
+    if (!isPipelineRunning) {
+      isPipelineRunning = true
+      runNextTask()
+    }
+  }
+
+  private fun waitForRemoteUpdate() {
+    val remoteLoadStatus = delegate.getRemoteLoadStatus()
+    if (remoteLoadStatus == ErrorRecoveryDelegate.RemoteLoadStatus.NEW_UPDATE_LOADED) {
+      runNextTask()
+    } else if (remoteLoadStatus == ErrorRecoveryDelegate.RemoteLoadStatus.NEW_UPDATE_LOADING ||
+      delegate.getCheckAutomaticallyConfiguration() != UpdatesConfiguration.CheckAutomaticallyConfiguration.NEVER) {
+      isWaitingForRemoteUpdate = true
+      if (delegate.getRemoteLoadStatus() != ErrorRecoveryDelegate.RemoteLoadStatus.NEW_UPDATE_LOADING) {
+        delegate.loadRemoteUpdate()
+      }
+      postDelayed({ handleRemoteLoadStatusChanged(ErrorRecoveryDelegate.RemoteLoadStatus.IDLE) }, REMOTE_LOAD_TIMEOUT_MS)
+    } else {
+      // there's no remote update, so move to the next step in the pipeline
+      pipeline.remove(Task.LAUNCH_NEW_UPDATE)
+      runNextTask()
+    }
+  }
+
+  private fun tryRelaunchFromCache() {
+    delegate.relaunch(object : Launcher.LauncherCallback {
+      override fun onFailure(e: Exception) {
+        // post to our looper, in case we're on a different thread now
+        post {
+          encounteredErrors.add(e)
+          pipeline.removeAll(setOf(Task.LAUNCH_NEW_UPDATE, Task.LAUNCH_CACHED_UPDATE))
+          runNextTask()
+        }
+      }
+
+      override fun onSuccess() {
+        // post to our looper, in case we're on a different thread now
+        post {
+          isPipelineRunning = false
+        }
```

Comment:
I'd be cautious of having this class shut down the looper since it's passed into the constructor. Conceptually, if ownership of the looper belongs to this class, then we should shut down the looper here, but otherwise the code that starts the looper probably should also stop it.

Reply:
I've added a call to `handlerThread.quitSafely()` in the outer ErrorRecovery class, instead, after the error handler is unregistered: https://github.com/expo/expo/pull/15220/commits/c3518a1f9d5ebbdd2fbe43613f00e4d83879cf81#diff-e52efeb214dc970084cc6f093b1102d318451cd43f31f9d15dd795f720b47530R99